Python/Product/VSInterpreters/PipPackageManager.cs

```diff
@@ -0,0 +1,651 @@
+                    } catch (OperationCanceledException) {
+                        IsReady = false;
+                    }
+                }
+
+                ui?.OnOutputTextReceived(IsReady ? Strings.InstallingPipSuccess : Strings.InstallingPackageFailed);
+                ui?.OnOperationFinished(operation, IsReady);
+            }
+        }
+
+        public async Task<bool> ExecuteAsync(string arguments, IPackageManagerUI ui, CancellationToken cancellationToken) {
+            AbortOnInvalidConfiguration();
+            await AbortIfNotReady(cancellationToken);
+
+            using (await _working.LockAsync(cancellationToken)) {
+                bool success = false;
+                string args;
+
+                if (!SupportsDashMPip) {
+                    args = "-c \"import pip; pip.main()\" ";
+                } else {
+                    args = "-m pip ";
+                }
+
+                args += arguments;
+
+                var operation = args;
+                ui?.OnOutputTextReceived(operation);
+                ui?.OnOperationStarted(Strings.ExecutingCommandStarted.FormatUI(arguments));
+
+                try {
+                    using (var output = ProcessOutput.Run(
+                        _factory.Configuration.InterpreterPath,
+                        new[] { args },
+                        _factory.Configuration.PrefixPath,
+                        UnbufferedEnv,
+                        false,
+                        PackageManagerUIRedirector.Get(ui),
+                        quoteArgs: false,
+                        elevate: await ShouldElevate(ui, operation)
+                    )) {
+                        if (!output.IsStarted) {
+                            return false;
+                        }
+                        var exitCode = await output;
+                        success = exitCode == 0;
+                    }
+                    return success;
+                } catch (IOException) {
+                    return false;
+                } finally {
+                    if (!success) {
+                        // Check whether we failed because pip is missing
+                        UpdateIsReadyAsync(true, CancellationToken.None).DoNotWait();
+                    }
+
+                    var msg = success ? Strings.ExecutingCommandSucceeded : Strings.ExecutingCommandFailed;
+                    ui?.OnOutputTextReceived(msg.FormatUI(arguments));
+                    ui?.OnOperationFinished(operation, success);
+                    await CacheInstalledPackagesAsync(true, cancellationToken);
+                }
+            }
+        }
+
+        public async Task<bool> InstallAsync(PackageSpec package, IPackageManagerUI ui, CancellationToken cancellationToken) {
+            AbortOnInvalidConfiguration();
+            await AbortIfNotReady(cancellationToken);
+
+            bool success = false;
+            List<string> args;
+
+            if (!SupportsDashMPip) {
```

Comment:
Technically "pip.main()" is not a supported entry point, so if they do something to break it I'd rather not break absolutely everywhere.

---
In reply to: [75519467](https://github.com/Microsoft/PTVS/pull/1536#discussion_r75519467) [](ancestors = 75519467)